### PR TITLE
[Distribute] Add distribute will exit callback

### PR DESCRIPTION
**Pull request by DmitriyKirakosyan**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* ~[ ] Did you check UI tests on the sample app? They are not executed on CI.~
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

This PR adds `distributeWillClose` callback to MSACDistributeDelegate which is called before the app gets terminated for an update installation.

## Related PRs or issues

[AB#84063](https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/84063)

### CODE REVIEW

File: AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h [hunk 1]
```diff
@@ -28,6 +28,14 @@ NS_SWIFT_NAME(DistributeDelegate)
  */
 - (BOOL)distribute:(MSACDistribute *)distribute releaseAvailableWithDetails:(MSACReleaseDetails *)details;
 
+/**
+ * Callback method that will be called before the app is permanently closed for update.
+ * It is the right place to add any required clean ups.
+ *
+ * @param distribute The instance of MSACDistribute.
+ */
+- (void)distributeWillClose:(MSACDistribute *)distribute;
```
Comment (iamclement) [inline AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h:37]: I am curious about why "close"? I suppose something more radical in this case could be more adequate like "terminate". Closing for me may convey that even though the UI closes the app may still be running in background. Just my thoughts.

Comment (iamclement) [reply AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h:37]: Or maybe "Terminate" might be confusing the OS lifcycle callback. "Exit" same as the HA name: https://github.com/bitstadium/HockeySDK-iOS/blob/develop/Classes/BITUpdateManagerDelegate.h#L158 is an option.

File: AppCenterDistribute/AppCenterDistributeTests/MSACDistributeTests.m [hunk 2]
```diff
@@ -2296,6 +2296,71 @@ - (void)testSetDelegate {
   XCTAssertEqual(strongDelegate, delegateMock);
 }
 
+- (void)testDelegateMethodWillCloseIsCalled {
+
+  // If
+  XCTestExpectation *willCloseIsCalledExpectation = [self expectationWithDescription:@"willCloseCalled"];
+
```
Comment (AnastasiaKubova) [inline AppCenterDistribute/AppCenterDistributeTests/MSACDistributeTests.m:2303]: empty lines

File: Sasquatch/SasquatchObjC/AppDelegate.m [hunk 3]
```diff
@@ -381,6 +381,17 @@ - (BOOL)distribute:(MSACDistribute *)distribute releaseAvailableWithDetails:(MSA
...
 }
 
+- (void)distributeWillClose:(MSACDistribute *)distribute {
+  dispatch_async(dispatch_get_main_queue(), ^{
+    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:NSLocalizedStringFromTable(@"distribute_alert_willclose_title", @"Sasquatch", @"")
+                                                                             message:NSLocalizedStringFromTable(@"distribute_alert_willclose_message", @"Sasquatch", @"")
+                                                                      preferredStyle:UIAlertControllerStyleAlert];
+    [self.window.rootViewController presentViewController:alertController animated:YES completion:nil];
+  });
+
```
Comment (AnastasiaKubova) [inline Sasquatch/SasquatchObjC/AppDelegate.m:391]: empty line

File: Sasquatch/SasquatchSwift/AppDelegate.swift [hunk 4]
```diff
@@ -315,6 +315,21 @@ extension AppDelegate: DistributeDelegate {
     }
     return false
   }
+
+  func distributeWillClose(_ distribute: Distribute) {
+    print("distributeWillClose callback invoked");
+
```
Comment (AnastasiaKubova) [inline Sasquatch/SasquatchSwift/AppDelegate.swift:321]: and here

File: AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h (hunk 1 above)
Comment (DmitriyKirakosyan) [reply AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h:37]: Makes sense. Let's name it `distributeWillExitApp`, the same as it's in the HA SDK?

**Review by AnastasiaKubova — APPROVED** [review]

**Review by AnastasiaKubova — APPROVED** [review]

**Review by Jamminroot — APPROVED** [review]

**Pull request merged**